contracts/finance/payment-splitter/traits.rs

```diff
@@ -0,0 +1,201 @@
+pub struct PaymentSplitterData {
+    pub total_shares: Balance,
+    pub total_released: Balance,
+    pub shares: StorageHashMap<AccountId, Balance>,
+    pub released: StorageHashMap<AccountId, Balance>,
+    pub payees: StorageVec<AccountId>,
+}
+
+declare_storage_trait!(PaymentSplitterStorage, PaymentSplitterData);
+
+#[derive(strum_macros::AsRefStr)]
+pub enum PaymentSplitterError {
+    LengthMismatch,
+    NoPayees,
+    AccountHasNoShares,
+    AccountIsNotDuePayment,
+    AccountZeroAddress,
+    SharesAreZero,
+    AlreadyHasShares,
+    TransferFailed,
+}
+
+/// This contract allows to split native token payments among a group of accounts. The sender does not need to be aware
+/// that the native token will be split in this way, since it is handled transparently by the contract.
+///
+/// The split can be in equal parts or in any other arbitrary proportion. The way this is specified is by assigning each
+/// account to a number of shares. Of all the native tokens that this contract receives, each account will then be able to claim
+/// an amount proportional to the percentage of total shares they were assigned.
+///
+/// PaymentSplitter `follows a _pull payment_ model. This means that payments are not automatically forwarded to the
+/// accounts but kept in this contract, and the actual transfer is triggered as a separate step by calling the `release`
+/// function.
+///
+/// ** Note **: In the substrate balance of contract decreases each block. Because it pays rent for the storage.
+/// So during `release`, each next user will get fewer native tokens.
+///
+/// This module is used through embedding of `PaymentSplitterData` and implementation of `PaymentSplitter` and
+/// `PaymentSplitterStorage` traits.
+#[brush::trait_definition]
+pub trait PaymentSplitter: PaymentSplitterStorage {
+    /// Getter for the total shares held by payees.
+    #[ink(message)]
+    fn total_shares(&self) -> Balance {
+        self.get().total_shares.clone()
+    }
+
+    /// Getter for the total amount of native token already released.
+    #[ink(message)]
+    fn total_released(&self) -> Balance {
+        self.get().total_released.clone()
+    }
+
+    /// Getter for the amount of shares held by an account.
+    #[ink(message)]
+    fn shares(&self, account: AccountId) -> Balance {
+        self.get().shares.get(&account).cloned().unwrap_or(0)
+    }
+
+    /// Getter for the amount of native token already released to a payee.
+    #[ink(message)]
+    fn released(&self, account: AccountId) -> Balance {
+        self.get().released.get(&account).cloned().unwrap_or(0)
+    }
+
+    /// Getter for the address of the payee number `index`.
+    #[ink(message)]
+    fn payee(&self, index: u32) -> AccountId {
+        self.get().payees.get(index).cloned().unwrap_or(ZERO_ADDRESS.into())
+    }
+
+    /// The native token received will be logged with `PaymentReceived` events. Note that these events are not fully
```

Comment:
The native token received will be logged with `PaymentReceived` events. Note that these events are not fully reliable: a contract can receive a native token without triggering this function. This only affects the reliability of the events and not the actual splitting of the native token.